File: r-package/R/remove_islands.R
#' Remove islands from Brazil
#'
#' @description
#' Removes Brazilian islands that are approximately more than 20 km from the
#' mainland coast. This is useful when analyses or data visualization should
#' focus on the continental territory of Brazil and exclude distant oceanic
#' islands.
#'
#' @param x An 'sf' object with CRS EPSG:4674. Usually an object returned from
#'        other geobr functions, such as `read_country()`, `read_states()`,
#'        `read_municipality()`, or similar functions.
#'
#' @return An `sf` data frame with the same attributes as `x`, but with distant
#'   islands removed from the geometry.
#'
#' @export
#'
#' @examples
#' library(geobr)
#' library(sf)
#'
#' br <- read_country(year=2022)
#'
#' br_no_islands <- remove_islands(br)
#'
#' plot(br)
remove_islands <- function(x){

  # Check input class
  checkmate::assert_class(x, classes = "sf")

  # Check input CRS
  x_crs <- sf::st_crs(x)

  if (!identical(x_crs$epsg, 4674L)) {
    cli::cli_abort(
      "{.arg x} must have CRS EPSG:4674 / SIRGAS 2000."
    )
  }

  # Path to simplified offshore buffer of Brazil, approximately 20 km from shore
  br_offcoast_path <- system.file(
    "extdata/br_offcoast.parquet",
    package = "geobr",
    mustWork = TRUE
  )

  br_offcoast <- br_offcoast_path |>
    arrow::open_dataset() |>
    sf::st_as_sf()

  # Fix eventual invalid geometries from input
  x <- duckspatial::ddbs_make_valid(x)

  # remove islands
  no_islands <- duckspatial::ddbs_difference(
    x = x,
    y = br_offcoast
    ) |>
    duckspatial::ddbs_collect()

  return(no_islands)
}

File: python-package/geobr/remove_islands.py
"""Remove distant Brazilian islands from spatial objects."""

from __future__ import annotations

from pathlib import Path

import geopandas as gpd


def _offcoast_path() -> Path:
    return Path(__file__).parent / "data" / "br_offcoast.parquet"


def remove_islands(x: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
    """Remove islands more than ~20 km from the mainland coast.

    Parameters
    ----------
    x : geopandas.GeoDataFrame
        Input with CRS EPSG:4674 (SIRGAS 2000).

    Returns
    -------
    geopandas.GeoDataFrame
        Same attributes with distant islands removed from geometry.
    """
    if not isinstance(x, gpd.GeoDataFrame):
        raise TypeError("`x` must be a geopandas.GeoDataFrame.")

    if x.crs is None or x.crs.to_epsg() != 4674:
        raise ValueError("`x` must have CRS EPSG:4674 / SIRGAS 2000.")

    offcoast = gpd.read_parquet(_offcoast_path())
    if offcoast.crs is None or offcoast.crs.to_epsg() != 4674:
        offcoast = offcoast.set_crs(4674)
    
    x = x.copy()
    x["geometry"] = x.geometry.make_valid()
    result = gpd.overlay(x, offcoast, how="difference")
    return result
